error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/OptionalOrElseTest.java

```diff
@@ -0,0 +1,135 @@
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(string + \"constant\");",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(\"constant\".toString());",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(string.toString());",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(this.string.toString());",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(String.valueOf(42));",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(string.toString().length());",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(\"constant\".equals(string));",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(string.equals(string));",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(this.string.equals(string));",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(foo());",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(this.foo());",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(new Object() {});",
+            "    // BUG: Diagnostic contains:",
+            "    optional.orElse(new int[0].length);",
+            "  }",
+            "",
+            "  private <T> T foo() {",
+            "    return null;",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  void replacement() {
+    BugCheckerRefactoringTestHelper.newInstance(OptionalOrElse.class, getClass())
+        .addInputLines(
+            "A.java",
+            "import java.util.Optional;",
+            "",
+            "class A {",
+            "  private final Optional<Object> optional = Optional.empty();",
+            "  private final String string = optional.toString();",
+            "",
+            "  void m() {",
+            "    optional.orElse(string + \"constant\");",
+            "    optional.orElse(\"constant\".toString());",
+            "    optional.orElse(string.toString());",
+            "    optional.orElse(this.string.toString());",
+            "    optional.orElse(String.valueOf(42));",
+            "    optional.orElse(string.toString().length());",
+            "    optional.orElse(string.equals(string));",
+            "    optional.orElse(foo());",
+            "    optional.orElse(this.<Number>foo());",
+            "    optional.orElse(this.<String, Integer>bar());",
+            "    optional.orElse(new Object() {});",
+            "    optional.orElse(new int[0].length);",
+            "  }",
+            "",
+            "  private <T> T foo() {",
+            "    return null;",
+            "  }",
+            "",
+            "  private <S, T> T bar() {",
+            "    return null;",
+            "  }",
+            "}")
```

Comment:
If I do this 
```suggestion
            "    optional.orElseGet(() -> new int[0].length);",
            "    optional.orElseGet(() -> B.NUMBER);",
            "    optional.orElseGet(B::getNumber);",
            "  }",
            "",
            "  private <T> T foo() {",
            "    return null;",
            "  }",
            "",
            "  private <S, T> T bar() {",
            "    return null;",
            "  }",
            "",
            "  class B {",
            "    static final int NUMBER = 1;",
            "",
            "    static String getNumber() {",
            "      return \"number\";",
            "    }",
            "  }",
            "}")
```

Shouldn't we expect `optional.orElseGet(() -> B.NUMBER)` to be `optional.orElseGet(() -> B.NUMBER)` ?

This is what I get instead 
![Screenshot 2024-02-16 at 17 03 47](https://github.com/PicnicSupermarket/error-prone-support/assets/110535847/33f7a4bf-3b71-49ab-8372-0439ba3fe1e4)

Reply:
This is intended :). The idea here is that field dereferencing (`foo.bar.baz` etc.) is _very_ efficient (conceptually: just some pointer dereferencing), so there's no point in deferring that work.